openest/generate/smart_curve.py

```diff
@@ -179,6 +179,80 @@ def format(self, lang):
+        if descriptions is None:
+            descriptions = {}
+        self.descriptions = descriptions
+
+        self.getters = [(lambda ds: ds._variables[variable]) if isinstance(variable, str) else variable for variable in self.variables] # functions return vector of length T
+
+    def __call__(self, ds):
+        maxtime = self.coeffmat.shape[1]
+        lindata = self.getters[0](ds)._data[:maxtime]
+        result = np.sum(self.coeffmat[0, :len(lindata)] * lindata)
+        
+        for ii in range(1, len(self.variables)):
+            if not self.allow_raising or self.variables[ii] in ds._variables:
+                termdata = self.getters[ii](ds)._data[:maxtime]
+                result += np.sum(self.coeffmat[ii, :len(lindata)] * termdata) # throws error if length mismatch
+            else:
+                result += np.sum(self.coeffmat[ii, :len(lindata)] * (lindata ** (ii + 1)))
+                
+        return result
+
+    def get_univariate(self):
+        raise NotImplementedError("Probably want to define a matrix-taking curve before this.")
+
+    def format(self, lang):
+        coeffvar = formatting.get_variable()
+        variable = formatting.get_variable()
+        funcvars = {}
+
+        repterms = []
+        if lang == 'latex':
+            if isinstance(self.variables[0], str):
+                repterms.append(r"%s_1 \cdot %s" % (coeffvar, variable))
+            else:
+                funcvar = formatting.get_function()
+                funcvars[self.variables[0]] = funcvar
+                repterms.append(r"%s_1 \cdot %s(%s)" % (coeffvar, funcvar, variable))
+        elif lang == 'julia':
+            if isinstance(self.variables[0], str):
+                repterms.append(r"sum(%s[1,:] * %s)" % (coeffvar, variable))
+            else:
+                funcvar = formatting.get_function()
+                funcvars[self.variables[0]] = funcvar
+                repterms.append(r"sum(%s[1,:] * %s(%s))" % (coeffvar, funcvar, variable))
+        
+        for ii in range(1, len(self.variables)):
+            if lang == 'latex':
+                if isinstance(self.variables[0], str):
+                    repterms.append(r"%s_1 \cdot %s^%d" % (coeffvar, variable, ii + 1))
+                else:
+                    funcvar = formatting.get_function()
+                    funcvars[self.variables[ii]] = funcvar
+                    repterms.append(r"%s_1 \cdot %s(%s)^%d" % (coeffvar, funcvar, variable, ii + 1))
+            elif lang == 'julia':
+                if isinstance(self.variables[0], str):
+                    repterms.append(r"sum(%s[1,:] * %s^%d)" % (coeffvar, variable, ii + 1))
+                else:
+                    funcvar = formatting.get_function()
+                    funcvars[self.variables[ii]] = funcvar
+                    repterms.append(r"sum(%s[1,:] * %s(%s)^%d)" % (coeffvar, funcvar, variable, ii + 1))
+
+        result = {'main': FormatElement(' + '.join(repterms))}
```

Comment:
Ignore me if I'm wrong but I don't think we import `FormatElement`, so unresolved reference here.